File: Algo/Medium/0238_Product_of_Array_Except_Self/Solution.py
from typing import *

## 2021-11-28
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        lr = [1] * len(nums)
        rl = [1] * len(nums)
        
        for i in range(0, len(nums)-1):
            lr[i+1] = lr[i] * nums[i]
        for j in range(len(nums)-1, 0, -1):
            rl[j-1] = rl[j] * nums[j]
                
        output = []
        for i in range(0, len(nums)):
            output.append(lr[i]*rl[i])
        
        return output


## 2022-08-18 
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        output = [1 for _ in nums]
        prefix_prod = 1
        for i in range(len(nums)):
            output[i] *= prefix_prod
            prefix_prod *= nums[i]
        
        prefix_prod = 1
        for i in range(len(nums)-1, -1, -1):
            output[i] *= prefix_prod
            prefix_prod *= nums[i]
        
        return output

File: Algo/Medium/0238_Product_of_Array_Except_Self/Solution.java
import java.util.*;

// 2021-05-11
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] output = new int[nums.length];
        int prod = 1;
        int zeroes = 0;
        int zeroindex = 0;
        
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {
                zeroes++;
                zeroindex = i;
            } else {
                prod *= nums[i];
            }
        }
        
        Arrays.fill(output, 0);
        
        if (zeroes > 1) return output;
        
        if (zeroes == 1) {
            output[zeroindex] = prod;
            return output;
        }
        
        for (int n = 0; n < output.length; n++) {
            output[n] = prod / nums[n];
        }
        return output;
    }
}

// 2021-06-29
class Solution2 {
    public int[] productExceptSelf(int[] nums) {
        int[] output = new int[nums.length];
        int total = 1;
        
        for (int right = 0; right < nums.length; right++) {
            output[right] = total;
            total *= nums[right];
        }
        
        total = 1;
        for (int left = nums.length - 1; left >= 0; left--) {
            output[left] *= total;
            total *= nums[left];
        }
        
        return output;
    }
}